### ISSUE TITLE
Use primitives for getChainStartIndices in MonotoneChainBuilder and MonotoneChainIndexer?

### ISSUE THREAD

**Issue by dr-jts** [locationtech/jts#463]
This brings forward [SF JTS issue 55](https://sourceforge.net/p/jts-topo-suite/bugs/55/).

@jandam do you have an improvement for this?

**Comment by jandam** [comment]
Proposed changes to MonotoneChainBuilder:

*) remove 
`public static int[] toIntArray(List list)
`

*) modify getChainStartIndices 
- replace ArrayList with integer array
```
/**
   * Return an array containing lists of start/end indexes of the monotone chains
   * for the given list of coordinates.
   * The last entry in the array points to the end point of the point array,
   * for use as a sentinel.
   */
  public static int[] getChainStartIndices(Coordinate[] pts)
  {
    // find the startpoint (and endpoints) of all monotone chains in this edge
    int start = 0;
    int[] startIndex = new int[Math.min(pts.length + 1, 16)];
    int dstIndex = 1;
    startIndex[0] = start;
    do {
      int last = findChainEnd(pts, start);
      if (dstIndex == startIndex.length) startIndex = Arrays.copyOf(startIndex, dstIndex + (dstIndex >> 1));
      startIndex[dstIndex++] = last;
      start = last;
    } while (start < pts.length - 1);

    if (dstIndex < startIndex.length) startIndex = Arrays.copyOf(startIndex, dstIndex);
    return startIndex;
  }
```
*) getChains 
- can use generics + ArrayList preallocation
```
public static List<MonotoneChain> getChains(Coordinate[] pts)
  {
    return getChains(pts, null);
  }

  /**
   * Return a list of the {@link MonotoneChain}s
   * for the given list of coordinates.
   */
  public static List<MonotoneChain> getChains(Coordinate[] pts, Object context)
  {
    int[] startIndex = getChainStartIndices(pts);
    List<MonotoneChain> mcList = new ArrayList<>(startIndex.length);
    for (int i = 0; i < startIndex.length - 1; i++) {
      MonotoneChain mc = new MonotoneChain(pts, startIndex[i], startIndex[i + 1], context);
      mcList.add(mc);
    }
    return mcList;
  }
```

**Comment by jandam** [comment]
List<int> is not supported by current versions of Java. There is some progress in 'Valhalla' project [1]. 
  I can contribute simple IntArrayList implementation that I use in my projects. There are also 3rd party primitive collections libraries but I prefer JTS without dependencies.
  Some basic information about collections can be found in [2] (randomly found by unnamed search service).

  - there should be no significant difference for small arrays between ArrayList<Integer> and IntArrayList
  - small (< 128) Integer instances should be cached
  - IntArrayList is better for huge arrays 
  - only 4 bytes / entry compared to ~20 bytes for Integer (just estimation)
  - better cache utilization - ints are allocated in continuous memory block
  - ArrayList<Integer> has internal Object array and every entry points to different part of memory (Integer instance)
  - JTS is used for processing huge GIS data (Spark extensions, ..) because it is great library
  - today server has huge amount of running parallel threads and today GCs are not so good for freeing unused instances.
  - sometimes I have to explicitly clear ArrayList because G1GC is not able to free its data in minor cycles. ArrayList instance is freed but internal array stays allocated and holding internal data.

[1] [Valhalla](https://wiki.openjdk.java.net/display/valhalla/) - https://wiki.openjdk.java.net/display/valhalla/
[2] [Empirical study of usage and performance of javacollections](https://ink.library.smu.edu.sg/cgi/viewcontent.cgi?article=4766&context=sis_research) - https://ink.library.smu.edu.sg/cgi/viewcontent.cgi?article=4766&context=sis_research

**Comment by dr-jts** [comment]
I implemented a simple `IntArrayList` and wired it into `MonotoneChainBuilder` in [this branch](https://github.com/dr-jts/jts/tree/try-mcbuilder-int-array).

But I don't see any significant performance difference (when running a large unary union, which should exercise this code fairly well).  

@jandam  Any thoughts?  Perhaps this just isn't a significant percentage of overall time?  If there's no real improvement I'm tempted just to keep using the built-in Java structures.

**Comment by dr-jts** [comment]
It seems like another option would be to avoid constructing the array of start indices entirely , and simply create the `MonotoneChain` objects during the first scan of the input points.

**Comment by dr-jts** [comment]
I created a patch to remove the use of integer arrays altogether in `MonotoneChainBuilder`.  See [this branch](https://github.com/dr-jts/jts/tree/improve-mcbuilder).

@jandam If you are able to review before I merge this that would be great.

**Comment by dr-jts** [comment]
I noticed that [MonotoneChainIndexer](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geomgraph/index/MonotoneChainIndexer.java) also has this issue, and in this case it's not possible to eliminate the int array.  So I will add the `IntArrayList` class to see what difference that makes.

**Comment by dr-jts** [comment]
See PR #467

**Comment by jandam** [comment]
I looked at `MonotoneChainBuilder.` 
getChains isn't protected from invalid input (pts.length <= 1).
 I suggest to return `Collections.emtyList()` or `throw IllegalArgumentException(...)`

IntArrayList:
```
  public void add(final int value) {
    ensureCapacity(size + 1);
    data[size] = value;
    ++size;
  }

  public void ensureCapacity(int capacity) {
      if (capacity <= data.length) return;
      capacity = Math.max(capacity, data.length * 2);
      data = Arrays.copyOf(data, capacity);
  }
```
I prefer above way to use ensureCapacity. Because resize logic is encapsulated inside. Where can be protected from int overflow.

Another example when you need addAll.

```
  public void addAll(final int[] values) {
   Objects.requireNonNull(values, "values");
   ensureCapacity(size + values.length);
   System.arraycopy(data, size, values, 0, values.length)
   size += values.length;
  }


```
Is there any reason not to use `Arrays.copyOf` ?

**Comment by jandam** [comment]
Personally I preffer following pattern for constructors. I call `this` with default arguments. It acts as optional arguments from other languages.
```
  public IntArrayList() {
    this(10);
  }

  public IntArrayList(final int initialCapacity) {
    data = new int[initialCapacity];
  }
```

**Comment by dr-jts** [comment]
> Personally I prefer following pattern for constructors. I call `this` with default arguments. It acts as optional arguments from other languages.

Yes, agreed.

**Comment by dr-jts** [comment]
> I prefer above way to use ensureCapacity. Because resize logic is encapsulated inside. Where can be protected from int overflow.
 
> Is there any reason not to use `Arrays.copyOf` ?

Excellent suggestions, thanks.

**Comment by dr-jts** [comment]
> Another example when you need addAll.
> 
> ```
>   public void addAll(final int[] values) {
>    Objects.requireNonNull(values, "values");
>    ensureCapacity(size + values.length);
>    System.arraycopy(data, size, values, 0, values.length)
>    size += values.length;
>   }
> ```

Thanks, good addition.  Note that I fixed a bug in this code (inputs need to be reversed in `arraycopy`)

**Comment by dr-jts** [comment]
Closed by #467
